File: app/src/main/java/dev/ragnarok/fenrir/util/existfile/FileExistJVM.java
package dev.ragnarok.fenrir.util.existfile;

import static dev.ragnarok.fenrir.util.Objects.nonNull;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.stream.JsonReader;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;

import dev.ragnarok.fenrir.model.Photo;
import dev.ragnarok.fenrir.model.wrappers.SelectablePhotoWrapper;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.AppPerms;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.core.Completable;

public class FileExistJVM implements AbsFileExist {
    private final List<String> CachedAudios = new LinkedList<>();
    private final List<String> RemoteAudios = new LinkedList<>();
    private final List<String> CachedPhotos = new LinkedList<>();
    private final Object isBusyLock = new Object();
    private boolean isBusy;

    private boolean setBusy(boolean nBusy) {
        synchronized (isBusyLock) {
            if (isBusy && nBusy) {
                return false;
            }
            isBusy = nBusy;
        }
        return true;
    }

    private void findRemoteAudios(Context context, boolean needLock) throws IOException {
        if (needLock) {
            if (!AppPerms.hasReadStoragePermissionSimple(context))
                return;
            if (!setBusy(true)) {
                return;
            }
        }
        RemoteAudios.clear();
        File audios = new File(Settings.get().other().getMusicDir(), "local_server_audio_list.json");
        if (!audios.exists()) {
            if (needLock) {
                setBusy(false);
            }
            return;
        }
        JsonReader reader = new JsonReader(new InputStreamReader(new FileInputStream(audios), StandardCharsets.UTF_8));
        reader.beginArray();
        while (reader.hasNext()) {
            RemoteAudios.add(reader.nextString().toLowerCase(Locale.getDefault()));
        }
        if (needLock) {
            setBusy(false);
        }
    }

    @Override
    public void findRemoteAudios(Context context) throws IOException {
        findRemoteAudios(context, true);
    }

    private String transform_owner(int owner_id) {
        if (owner_id < 0)
            return "club" + Math.abs(owner_id);
        else
            return "id" + owner_id;
    }

    private boolean existPhoto(Photo photo) {
        for (String i : CachedPhotos) {
            if (i.contains(transform_owner(photo.getOwnerId()) + "_" + photo.getId())) {
                return true;
            }
        }
        return false;
    }

    private void loadDownloadPath(String Path) {
        File temp = new File(Path);
        if (!temp.exists())
            return;
        File[] file_list = temp.listFiles();
        if (file_list == null || file_list.length <= 0)
            return;
        for (File u : file_list) {
            if (u.isFile())
                CachedPhotos.add(u.getName());
            else if (u.isDirectory()) {
                loadDownloadPath(u.getAbsolutePath());
            }
        }
    }

    @Override
    public Completable findLocalImages(@Nullable List<SelectablePhotoWrapper> photos) {
        return Completable.create(t -> {
            if (!setBusy(true)) {
                return;
            }
            File temp = new File(Settings.get().other().getPhotoDir());
            if (!temp.exists()) {
                setBusy(false);
                t.onComplete();
                return;
            }
            File[] file_list = temp.listFiles();
            if (file_list == null || file_list.length <= 0) {
                setBusy(false);
                t.onComplete();
                return;
            }
            CachedPhotos.clear();
            for (File u : file_list) {
                if (u.isFile())
                    CachedPhotos.add(u.getName());
                else if (u.isDirectory()) {
                    loadDownloadPath(u.getAbsolutePath());
                }
            }
            if (nonNull(photos)) {
                for (SelectablePhotoWrapper i : photos) {
                    i.setDownloaded(existPhoto(i.getPhoto()));
                }
            }
            setBusy(false);
            t.onComplete();
        });
    }

    @Override
    public void addAudio(@NonNull String file) {
        if (!setBusy(true)) {
            return;
        }
        CachedAudios.add(file.toLowerCase(Locale.getDefault()));
        setBusy(false);
    }

    @Override
    public Completable findAllAudios(Context context) {
        if (!AppPerms.hasReadStoragePermissionSimple(context))
            return Completable.complete();
        return Completable.create(t -> {
            if (!setBusy(true)) {
                return;
            }
            findRemoteAudios(context, false);
            File temp = new File(Settings.get().other().getMusicDir());
            if (!temp.exists()) {
                setBusy(false);
                t.onComplete();
                return;
            }
            File[] file_list = temp.listFiles();
            if (file_list == null || file_list.length <= 0) {
                setBusy(false);
                t.onComplete();
                return;
            }
            CachedAudios.clear();
            for (File u : file_list) {
                if (u.isFile())
                    CachedAudios.add(u.getName().toLowerCase(Locale.getDefault()));
            }
            setBusy(false);
        });
    }

    @Override
    public void markExistPhotos(@NonNull List<SelectablePhotoWrapper> photos) {
        synchronized (isBusyLock) {
            if (isBusy) {
                return;
            }
            if (Utils.isEmpty(photos)) {
                return;
            }
            for (SelectablePhotoWrapper i : photos) {
                i.setDownloaded(existPhoto(i.getPhoto()));
            }
        }
    }

    @Override
    public boolean isExistRemoteAudio(@NonNull String file) {
        synchronized (isBusyLock) {
            if (isBusy) {
                return false;
            }
            String res = file.toLowerCase(Locale.getDefault());
            if (!Utils.isEmpty(RemoteAudios)) {
                for (String i : RemoteAudios) {
                    if (i.equals(res)) {
                        return true;
                    }
                }
            }
            return false;
        }
    }

    @Override
    public int isExistAllAudio(@NonNull String file) {
        synchronized (isBusyLock) {
            if (isBusy) {
                return 0;
            }
            String res = file.toLowerCase(Locale.getDefault());
            if (!Utils.isEmpty(CachedAudios)) {
                for (String i : CachedAudios) {
                    if (i.equals(res)) {
                        return 1;
                    }
                }
            }
            if (!Utils.isEmpty(RemoteAudios)) {
                for (String i : RemoteAudios) {
                    if (i.equals(res)) {
                        return 2;
                    }
                }
            }
            return 0;
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/util/existfile/FileExistJVM.kt
package dev.ragnarok.fenrir.util.existfile

import android.content.Context
import dev.ragnarok.fenrir.model.Photo
import dev.ragnarok.fenrir.model.wrappers.SelectablePhotoWrapper
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.AppPerms.hasReadStoragePermissionSimple
import dev.ragnarok.fenrir.util.serializeble.json.internal.JavaStreamSerialReader
import dev.ragnarok.fenrir.util.serializeble.json.internal.WriteMode
import dev.ragnarok.fenrir.util.serializeble.json.internal.lexer.ReaderJsonLexer
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.CompletableEmitter
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.util.*
import kotlin.math.abs

class FileExistJVM : AbsFileExist {
    private val CachedAudios: MutableList<String> = LinkedList()
    private val RemoteAudios: MutableList<String> = LinkedList()
    private val CachedPhotos: MutableList<String> = LinkedList()
    private val isBusyLock = Any()
    private var isBusy = false
    private fun setBusy(nBusy: Boolean): Boolean {
        synchronized(isBusyLock) {
            if (isBusy && nBusy) {
                return false
            }
            isBusy = nBusy
        }
        return true
    }

    @Throws(IOException::class)
    private fun findRemoteAudios(context: Context, needLock: Boolean) {
        if (needLock) {
            if (!hasReadStoragePermissionSimple(context)) return
            if (!setBusy(true)) {
                return
            }
        }
        RemoteAudios.clear()
        val audios = File(Settings.get().other().musicDir, "local_server_audio_list.json")
        if (!audios.exists()) {
            if (needLock) {
                setBusy(false)
            }
            return
        }
        try {
            val reader = ReaderJsonLexer(
                JavaStreamSerialReader(FileInputStream(audios))
            )
            reader.consumeNextToken(WriteMode.LIST.begin)
            while (reader.canConsumeValue()) {
                RemoteAudios.add(reader.consumeStringLenient().lowercase(Locale.getDefault()))
                reader.tryConsumeComma()
            }
            reader.consumeNextToken(WriteMode.LIST.end)
            if (needLock) {
                setBusy(false)
            }
        } catch (e: Exception) {
            if (needLock) {
                setBusy(false)
            }
            throw e
        }
    }

    @Throws(IOException::class)
    override fun findRemoteAudios(context: Context) {
        findRemoteAudios(context, true)
    }

    private fun transform_owner(owner_id: Int): String {
        return if (owner_id < 0) "club" + abs(owner_id) else "id$owner_id"
    }

    private fun existPhoto(photo: Photo): Boolean {
        for (i in CachedPhotos) {
            if (i.contains(transform_owner(photo.ownerId) + "_" + photo.getObjectId())) {
                return true
            }
        }
        return false
    }

    private fun loadDownloadPath(Path: String) {
        val temp = File(Path)
        if (!temp.exists()) return
        val file_list = temp.listFiles()
        if (file_list == null || file_list.isEmpty()) return
        for (u in file_list) {
            if (u.isFile) CachedPhotos.add(u.name) else if (u.isDirectory) {
                loadDownloadPath(u.absolutePath)
            }
        }
    }

    override fun findLocalImages(photos: List<SelectablePhotoWrapper>): Completable {
        return Completable.create { t: CompletableEmitter ->
            if (!setBusy(true)) {
                return@create
            }
            val temp = File(Settings.get().other().photoDir)
            if (!temp.exists()) {
                setBusy(false)
                t.onComplete()
                return@create
            }
            val file_list = temp.listFiles()
            if (file_list == null || file_list.isEmpty()) {
                setBusy(false)
                t.onComplete()
                return@create
            }
            CachedPhotos.clear()
            for (u in file_list) {
                if (u.isFile) CachedPhotos.add(u.name) else if (u.isDirectory) {
                    loadDownloadPath(u.absolutePath)
                }
            }
            for (i in photos) {
                i.setDownloaded(existPhoto(i.photo))
            }
            setBusy(false)
            t.onComplete()
        }
    }

    override fun addAudio(file: String) {
        if (!setBusy(true)) {
            return
        }
        CachedAudios.add(file.lowercase(Locale.getDefault()))
        setBusy(false)
    }

    override fun addPhoto(file: String) {
        if (!setBusy(true)) {
            return
        }
        CachedPhotos.add(file.lowercase(Locale.getDefault()))
        setBusy(false)
    }

    override fun findAllAudios(context: Context): Completable {
        return if (!hasReadStoragePermissionSimple(context)) Completable.complete() else Completable.create { t: CompletableEmitter ->
            if (!setBusy(true)) {
                t.onComplete()
                return@create
            }
            findRemoteAudios(context, false)
            val temp = File(Settings.get().other().musicDir)
            if (!temp.exists()) {
                setBusy(false)
                t.onComplete()
                return@create
            }
            val file_list = temp.listFiles()
            if (file_list == null || file_list.isEmpty()) {
                setBusy(false)
                t.onComplete()
                return@create
            }
            CachedAudios.clear()
            for (u in file_list) {
                if (u.isFile) CachedAudios.add(u.name.lowercase(Locale.getDefault()))
            }
            setBusy(false)
            t.onComplete()
        }
    }

    override fun markExistPhotos(photos: List<SelectablePhotoWrapper>) {
        synchronized(isBusyLock) {
            if (isBusy) {
                return
            }
            if (photos.isEmpty()) {
                return
            }
            for (i in photos) {
                i.setDownloaded(existPhoto(i.photo))
            }
        }
    }

    override fun isExistRemoteAudio(file: String): Boolean {
        synchronized(isBusyLock) {
            if (isBusy) {
                return false
            }
            val res = file.lowercase(Locale.getDefault())
            if (RemoteAudios.nonNullNoEmpty()) {
                for (i in RemoteAudios) {
                    if (i == res) {
                        return true
                    }
                }
            }
            return false
        }
    }

    override fun isExistAllAudio(file: String): Int {
        synchronized(isBusyLock) {
            if (isBusy) {
                return 0
            }
            val res = file.lowercase(Locale.getDefault())
            if (CachedAudios.nonNullNoEmpty()) {
                for (i in CachedAudios) {
                    if (i == res) {
                        return 1
                    }
                }
            }
            if (RemoteAudios.nonNullNoEmpty()) {
                for (i in RemoteAudios) {
                    if (i == res) {
                        return 2
                    }
                }
            }
            return 0
        }
    }
}